Build workspace packages in top-sorted order

JordanMartinez opened the pull request:
### Description of the change

First part of #988.

- When building all workspace packages, builds them in a reverse-topological-sort order (i.e. leaves first, then roots)
- Adds a duplicate module check when building all workspace packages

Tests added cover these cases:


```mermaid
flowchart TD
  subgraph "Case 1"
    A ---> Dep0["effect, console, prelude"]
    B ---> Dep0
  end
  
  subgraph "Case 2"
    A2 ---> effect2
    A2 ---> console2
    A2 ---> prelude2
    A2 ---> Shared2
    B2 ---> effect2
    B2 ---> console2
    B2 ---> prelude2
    B2 ---> Shared2
    Shared2 ---> prelude2
  end

  subgraph "Case 3"
    A3 ---> effect3
    A3 ---> console3
    A3 ---> prelude3
    A3 ---> B3
    A3 ---> C3
    B3 ---> effect3
    B3 ---> console3
    B3 ---> prelude3
    B3 ---> C3
    C3 ---> prelude3
  end

  subgraph "Case 4 (duplicate module)"
    A4 ---> effect4
    A4 ---> console4
    A4 ---> prelude4
    B4 ---> effect4
    B4 ---> console4
    B4 ---> prelude4
    C4 ---> prelude4
  end
```

### Checklist:

- [ ] Added the change to the "Unreleased" section of the changelog
- [ ] Added some example of the new feature to the `README`
- [ ] Added a test for the contribution (if applicable)

**P.S.**: the above checks are not compulsory to get a change merged, so you may skip them. However, taking care of them will result in less work for the maintainers and will be much appreciated 😊

JordanMartinez commented on bin/src/Main.purs:
```diff
@@ -819,10 +819,9 @@ mkReplEnv replArgs dependencies = do
-      Nothing -> Config.getWorkspacePackages workspace.packageSet
+  selected <- case workspace.selected of
+    Just s -> pure s
+    Nothing -> die "The REPL requires you to select a package"
```
Here's one potential downside of this change... But I guess this is just making what was an implicit problem explicit...?

f-f replied:
Do we have to change this?

JordanMartinez replied:
What happens when there are two modules with the same package name? I guess the only option would be to force the user to specify a package that exists only for the repl?

f-f replied:
You mean two packages with the same module name? We shouldn't allow that - we'll probably need to run a graph check when compiling more than one package so we can detect this.

JordanMartinez replied:
I've updated the code to reallow this, but haven't yet implemented the duplicate module check yet.

JordanMartinez commented:
The three cases I test for are similar to but not quite those described in the #988. Unless we publish a package that intentionally has a warning in it, it will not be possible to test for those cases specifically. So, the best we can do is test against local workspace packages that satisfy that criteria.

JordanMartinez commented on src/Spago/Command/Build.purs:
```diff
@@ -171,11 +179,48 @@ getBuildGlobs { selected, dependencies, withTests, depsOnly } =
+  dependencyGlobs = (Tuple.uncurry $ Config.sourceGlob NoTestGlobs) =<< dependencyPkgs
+  monorepoPkgGlobs
+    | depsOnly = []
+    | otherwise = (Tuple.uncurry $ Config.sourceGlob NoTestGlobs) =<< monorepoPkgs
```
CI is failing because of the changes I made here. The underlying issue is that the globs don't include the dependencies of the package's direct dependencies. I'm not as familiar with this part of the code, so @f-f could you give some insight as to how I should best handle this?

I basically want to write
```
Set.fromFoldable $ projectGlobs <> dependencyGlobs <> [ BuildInfo.buildInfoPath ]
```
but it seems that `dependencies` includes the dependencies for all workspace packages, not just the one for which I'm currently getting globs.

f-f replied:
Right, we collect all the dependencies needed in the `Spago.Command.Fetch` module - we could collect a different set of dependencies for each package there...

...however this is problematic as well: if you'd like package A to hide warnings from its dependencies, and package B to definitely show you warnings for the dependencies, then building package A first will mean that Spago will build its dependencies, hide the warnings, and keep a cache of this, so when building package B no warnings will be shown either.

A possible way out of this could be to treat the external dependencies of the project as a whole (i.e. as things are right now) - this would mean that we'd end up configuring each package PSA options separately (i.e. strict, hiding warnings, etc, but only for the code in the project), but configuring the behaviour of how to treat the warnings from dependencies at the workspace level instead.

JordanMartinez replied:
That sounds fine to me. Most people probably don't care about the library code anyways since it's beyond their control.

JordanMartinez commented:
Tasks remaining:
- [ ] Check for duplicate modules in the workspace packages and error if found
- [ ] Add a test for the above check

JordanMartinez commented on src/Spago/Command/Fetch.purs:
```diff
@@ -105,12 +137,23 @@ run { packages, ensureRanges, isTest } = do
   -- if the flag is selected, we kick off the process of adding ranges to the config
   when ensureRanges do
-    { configPath, yamlDoc } <- getPackageConfigPath
+    { configPath, yamlDoc, transitiveDeps } <- case workspaceDepsOrSelectedPackage of
+      Right r ->
+        pure r
+      Left _ ->
+        die
+          [ "No package found in the root configuration."
+          , "Please use the `-p` flag to select a package to which to add ranges."
+          ]
```
Is there a util for something like
```
foo `orDieWith`
  [ "error message" ]
```
where `orDieWith` is defined as
```purs
orDieWith :: forall a msg y. Maybe a -> msg -> Spago y a
orDieWith mb msg = case mb of
  Just a -> pure a
  Nothing -> die msg
```

JordanMartinez commented on src/Spago/Command/Publish.purs:
```diff
@@ -105,8 +106,9 @@ publish _args = do
     , git: env.git
     , purs: env.purs
-    , selected: env.selected
+    , selected: env.selected -- Note: should this be using the updated `selected` value?
```
Just highlighting this question...

f-f replied:
Actually yeah, let's use the new value

JordanMartinez commented:
@f-f For this code https://github.com/purescript/spago/blob/32614da384cc10e147ec658146dd59b16769d246/src/Spago/Purs/Graph.purs#L54-L56
could you clarify what the mappings are?

f-f commented:
Sure - let's add a comment to clarify it for future reference: for every Package that we depend on, we note which Module we are depending on, and for each of them we note from which Module we are importing it.

JordanMartinez commented:
K this should be good to go. Opening thread has been updated to include a visualization of the 4 cases being tested.

f-f commented:
Thanks! I'll have a look through it now

f-f commented on src/Spago/Purs/Graph.purs:
```diff
@@ -149,32 +165,39 @@ checkImports = do
+      pure $ Just graph
 
-      pure result
+runGraphCheck :: forall a. WorkspacePackage -> Set FilePath -> Array String -> Spago (PreGraphEnv a) (Array Docc)
```
We don't need this function if we are exposing `runGraph` and `toImportErrors`

JordanMartinez replied:
Oh ok. I wasn't sure if the API here was used somewhere downstream as `checkImports` was being exported but isn't used anywhere else AFAIK.

JordanMartinez replied:
Done.

f-f commented on src/Spago/Command/Build.purs:
```diff
@@ -140,42 +97,190 @@ run opts = do
+            toModulesDefinedByThisPackage :: Purs.ModuleGraphNode -> Maybe (Array (Tuple PackageName FilePath))
+            toModulesDefinedByThisPackage v =
+              [ Tuple selected.package.name v.path ] <$ (String.stripPrefix (String.Pattern selected.path) v.path)
+            modulesDefinedByThisPackage = Map.mapMaybe toModulesDefinedByThisPackage $ unwrap graph
```
I was entirely convinced that `purs graph` would fail with `DuplicateModule` in case of modules with the same name, but no, it just takes the last! Or at least I can't get it to fail, maybe I'm not trying hard enough.

f-f replied:
If we indeed have to do all this bookkeeping, then it should be moved to the Graph module, so we can have a nicely contained `detectDuplicateModules`, or something like that

f-f replied:
An alternative to doing the bookkeeping is of course to just build everything all together and let it fail with `DuplicateModule` from `purs`

JordanMartinez replied:
> I was entirely convinced that `purs graph` would fail with `DuplicateModule` in case of modules with the same name, but no, it just takes the last! Or at least I can't get it to fail, maybe I'm not trying hard enough.

I think you're running into https://github.com/purescript/purescript/issues/4342

f-f replied:
Oh no, what a nightmare. Do you think we should fix that now, or paper over it here?

JordanMartinez replied:
It's always better to fix it up-stream, but the problem is that `purs graph` doesn't work until the latest release. So, we'll still need to paper over it here.

f-f reviewed (commented):
Had a quick pass, I'll do another later/tomorrow

JordanMartinez commented on src/Spago/Command/Fetch.purs:
```diff
@@ -2,10 +2,12 @@ module Spago.Command.Fetch
   ( FetchEnv
   , FetchEnvRow
   , FetchOpts
+  , PackageDependencyInfo
```
AFAIK, the behavior is unchanged. There are only 2 possible states that can occur when producing the `transitiveDeps` value here:
1. a package was selected (whether it was the root package or some other package is irrelevant). In which case, we get dependencies only for that package.
2. no package was selected. In which case, we get dependencies for all packages.

But, I updated this code so that I am also recording which package corresponds with which `PackageMap` value when calling `getTransitiveDeps`. Otherwise, the `Build.getBuildGlobs`' `dependencies` argument includes a given package's dependencies plus the other workspace package dependencies. Using `selected.package.dependencies` in `Build.getBuildGlobs` returns only the direct dependencies of that package, not the transitive ones.

f-f replied:
I see - then I think it's hard for me to see the need for both fields in this record:
- if we have a map of dependencies for each package, then we can just take a union of all the values when we need to have all the dependencies, that sounds like we don't need to keep the `dependencies` field around?

JordanMartinez replied:
I originally kept it because it led to less code churn due to the renaming of the field and its corresponding type while I worked on other things. Moreover, calculating `dependencies` here once meant I didn't need to recalculate it elsewhere in every usage thereafter. But, on the other hand, that means two pieces of state that could diverge at some point. So, I've dropped the `dependencies` field.

JordanMartinez commented on src/Spago/Purs/Graph.purs:
```diff
@@ -51,8 +55,20 @@ type PackageGraphNode =
+-- | `Map.singleton "prelude" $ Map.singleton "Prelude" $ Set.singleton "MyModule"``
+-- |
+-- | Every package can be imported by several project modules.
+-- | In each project module, several modules from the same package can be imported.
```
I think I'll just drop the tail.

JordanMartinez commented:
Another idea came to mind this morning, but if we do implement it, I'd like to do so in a separate PR. Let's say I have three packages A, B, C, and D with the following dependency structure: C and D both depend on A and B whereas A and B are independent of one another:

```mermaid
flowchart TD
  C ---> A
  C ---> B
  D ---> A
  D ---> B
```
Let's assume that `C` will fail to build whereas `D` will succeed. If I run `spago build`, I would assume that the command will build as many packages as can be built. In other words, let's say we build `C` before `D`. Even though `C` fails to build, `D` should still be built. Right now, if `C` fails to build, I believe `D` won't even be attempted.

f-f commented:
> Even though `C` fails to build, `D` should still be built. Right now, if `C` fails to build, I believe `D` won't even be attempted.

Does it matter? The build is broken and needs fixing 🙂 

In all seriousness - we could attempt all the builds and give all the errors together, but not sure it would be very useful? It's often the case that one needs to go through the fix-compile-repeat loop many times as the compiler errors out as soon as it finds errors in the modules it's building and can't build their dependants.

JordanMartinez commented:
Good points. The user could also call `spago build -p` manually for the packages that did build, so this doesn't need to be implemented.

f-f commented on src/Spago/Config.purs:
```diff
@@ -453,6 +455,19 @@ getWorkspacePackages = Array.mapMaybe extractWorkspacePackage <<< Map.toUnfoldab
     Tuple _ (WorkspacePackage p) -> Just p
     _ -> Nothing
 
+getToplogicallySortedWorkspacePackages :: PackageSet -> Array WorkspacePackage
+getToplogicallySortedWorkspacePackages packageSet = do
```
```suggestion
getTopologicallySortedWorkspacePackages :: PackageSet -> Array WorkspacePackage
getTopologicallySortedWorkspacePackages packageSet = do
```

f-f commented on src/Spago/Command/Run.purs:
```diff
@@ -97,11 +97,11 @@ run = do
       let
         globs = Build.getBuildGlobs
-          { dependencies
+          { dependencies: foldl (Map.unionWith (\l _ -> l)) Map.empty packageDependencies
```
`Fetch.getAllDependencies` here too

f-f commented on src/Spago/Command/Build.purs:
```diff
@@ -140,42 +96,193 @@ run opts = do
+  , dependencies :: Map PackageName Package
+  }
+
+getEntireWorkspaceGlobs :: BuildWorkspaceGlobsOptions -> Set FilePath
```
Are these implementations different enough to warrant a split? It feels like we could cover everything with a few datatypes (I mean, the previous implementation was covering everything, so why not) - adding a different one means they can eventually diverge, which I'd rather not do

f-f reviewed (commented):
Had another round of reviewing - I don't seem to be able to wrap my head around the patch in one go, but we'll get there 😄

f-f commented:
Ah, I also tested it and it's pleasantly fast - I was a bit concerned that it would be slower to call purs so many times, but it's actually alright.

Something else that I noticed was that we show the PSA results for every package (and the `Build succeeded` message too) - I wonder if we should gather all the things in a single PSA display?

<img width="550" alt="image" src="https://github.com/purescript/spago/assets/5480361/b6c65914-65bb-4e7b-a7f8-36de80ad4676">

JordanMartinez commented on src/Spago/Command/Init.purs:
```diff
@@ -84,42 +88,51 @@ run opts = do
-        , backend: Nothing
-        , lock: Nothing
-        }
+  , workspace: (These.theseRight opts) <#> \{ setVersion } ->
```
Well, a config file is one of three values: 
- A package only 
- A workspace only 
- Both of them 

These summarizes this idea without allowing for am impossible state like two Nothings if it's represented using two Maybes

JordanMartinez replied:
Since each field here (i.e. package and workspace) are maybe values, I'm pulling out the corresponding value from the These value (if it exists).

f-f replied:
Oh I see now! I think we might be better off with a custom type that has three possible values - it's equivalent to `These`, but more self-explanatory and less surprising. We might want to do the same with the new `Either`s that are being added too

JordanMartinez commented on src/Spago/Command/Fetch.purs:
```diff
@@ -431,3 +459,21 @@ getVersionFromPackage :: Package -> Version
   _ -> unsafeFromRight $ Version.parse "0.0.0"
+
+-- TODO I guess this should be configurable
+replSupportPackageName :: PackageName
```
Should i put it in a module like Repl.SupportPackage?

f-f replied:
I guess we could have a `Spago.Repl` module with it

JordanMartinez commented:
Latest commits address some of the feedback, but I'll have to get to the rest of this on Monday.

JordanMartinez commented:
> Something else that I noticed was that we show the PSA results for every package (and the `Build succeeded` message too) - I wonder if we should gather all the things in a single PSA display?
> <img alt="image" width="550" src="https://user-images.githubusercontent.com/5480361/271772209-b6c65914-65bb-4e7b-a7f8-36de80ad4676.png">

I think this should be further discussed before we work on it. At the very least, I don't plan on doing anything for this in this PR.

JordanMartinez commented:
I believe I've addressed all feedback. I'm still running tests locally to verify that everything still passes.

f-f commented on bin/src/Main.purs:
```diff
@@ -515,7 +516,7 @@ main =
             { env, fetchOpts } <- mkFetchEnv { packages, selectedPackage, ensureRanges, testDeps }
             -- TODO: --no-fetch flag
-            dependencies <- runSpago env (Fetch.run fetchOpts)
+            packageDependencies <- runSpago env (Fetch.run fetchOpts)
```
```suggestion
            dependencies <- runSpago env (Fetch.run fetchOpts)
```

This is a bit of a nit, but after these few iterations I think it would make sense to go back to calling this `dependencies`, mostly for terseness, but also so that we can use `packageDependencies` for when we extract the set of dependencies from a package

JordanMartinez replied:
If `Map PackageName PackageMap` was defined as a type (preferably a newtype per my comment about hating type aliases below), then this value could be named after that type. Not sure what that name would be. Perhaps `packageToDependencies`? Verbose, yes, but still leaves room for `packageDependencies`, `dependencies`. Moreover, it implies a mapping of some sort.

f-f replied:
I think what I'm getting at is that the content carried by this name has not changed so much to warrant a project-wide rename

f-f commented on bin/src/Main.purs:
```diff
@@ -844,23 +845,23 @@ mkPublishEnv dependencies = do
-      Nothing -> Config.getWorkspacePackages workspace.packageSet
+  selected <- case workspace.selected of
+    Just s -> pure $ Build.SinglePackageGlobs s
+    Nothing -> pure $ Build.AllWorkspaceGlobs $ Config.getWorkspacePackages workspace.packageSet
```
I think `AllWorkspaceGlobs` always implies that we `getWorkspacePackages`, so we should call this then we go use the value rather than when we construct it (or use a smart constructor)
```suggestion
    Nothing -> pure $ Build.AllWorkspaceGlobs workspace.packageSet
```

f-f commented on src/Spago/Command/Publish.purs:
```diff
@@ -125,9 +126,12 @@ publish _args = do
-  let globs = Build.getBuildGlobs { selected: [ selected ], withTests: false, dependencies, depsOnly: false }
-  graphCheckErrors <- Graph.runGraphCheck selected globs []
-  for_ graphCheckErrors addError
+  let dependencies = Fetch.getAllDependencies packageDependencies
```
Looking at this diff reminds me that we need a dry-run kind of test for the publish command, there is nothing that ensures that it keeps working at the moment 😅

JordanMartinez replied:
So, is this you saying that we need to add a test for that inside or outside this PR? I'm hoping outside?

f-f replied:
Of course, definitely not here! I was just issuing a general reminder for both of us 😄

f-f commented on src/Spago/Command/Fetch.purs:
```diff
@@ -250,7 +297,10 @@ run { packages, ensureRanges, isTest } = do
-  pure transitivePackages
+  pure packageDependencies
+
+getAllDependencies :: Map PackageName PackageMap -> PackageMap
```
Uhhh I think I suggested this name, but `Fetch.getAllDependencies` has a "download things" sort of vibe. Not sure what a better name would be though, maybe just `allDependencies`? 🤔 

Then we could have a type alias `type AllDependencies = Map PackageName PackageMap` to make it extra clear

JordanMartinez replied:
Personally, I hate type aliases because they disappear in errors and because they more often hide the underlying structure that is informative. Would you be against defining this as a newtype?

f-f replied:
Newtypes are good for things that can be mixed up with other things, but I doubt we can misuse a `Map PackageName PackageMap` in the same way you'd misuse a `String`.

This is the only place where type aliases shine, because it doesn't really matter if the compiler will make it disappear, it's just there to hide the inner structure (which is a feature in this case. You'd only consume this type with the functions from this file)

JordanMartinez replied:
I went with `PackageTransitiveDeps` for a type alias and `toAllDependencies` due to working on the code before seeing this. Would you like me to use your naming?

f-f replied:
No, this is great!

f-f commented on src/Spago/Command/Fetch.purs:
```diff
@@ -33,18 +34,28 @@ import Registry.Sha256 as Sha256
+import Spago.Repl as Repl
 import Spago.Tar as Tar
 
+data PackageSelection
+  = SinglePackage
+      { workspacePackage :: WorkspacePackage
+      , configPath :: FilePath
+      , yamlDoc :: YamlDoc Core.Config
+      , transitiveDeps :: PackageMap
+      }
+  | AllWorkspacePackages (Map PackageName PackageMap)
+
```
There's a duality between this type above and this type defined in the `Build` module:

```
data SelectedPackageGlob
  = SinglePackageGlobs WorkspacePackage
  | AllWorkspaceGlobs (Array WorkspacePackage)
```

I wonder if it's worth collapsing everything into this one so we avoid the wrapping/unwrapping, something like
```suggestion
data PackageSelection
  = SinglePackage
      { workspacePackage :: WorkspacePackage
      , configPath :: FilePath
      , yamlDoc :: YamlDoc Core.Config
      , transitiveDeps :: PackageMap
      }
  | AllWorkspacePackages (Map WorkspacePackage PackageMap)

```

JordanMartinez replied:
Originally... In `Fetch`, the additional fields (i.e. `configPath` and `yamlDoc`) were added so we wouldn't need to construct them twice / do `case _ of` twice part when running the code under `when ensureRanges` and `when (not $ Array.null packages)`. In `Build`, this extra info isn't needed and may be hard to construct.

So, you might say, let's define a type:
```purs
data PackageSelection a b
  = SinglePackage a
  | AllWorkspacePackges b
```
the works across these two. I guess that works; this just feels super nitpicky to me.

f-f replied:
> In Build, this extra info isn't needed and may be hard to construct.

Ah you're right indeed - Build needs less stuff. Nevermind then!

f-f commented on src/Spago/Command/Fetch.purs:
```diff
@@ -109,12 +140,28 @@ run { packages, ensureRanges, isTest } = do
   -- if the flag is selected, we kick off the process of adding ranges to the config
   when ensureRanges do
-    { configPath, yamlDoc } <- getPackageConfigPath
+    { configPath, yamlDoc, transitiveDeps } <- case selectedPackage of
+      SinglePackage p -> pure p
+      AllWorkspacePackages _ -> die
+        [ "No package found in the root configuration."
+        , "Please use the `-p` flag to select a package to which to add ranges."
+        ]
```
This error message was packaged in a single function and is now repeated three times, could we reintroduce that?

JordanMartinez replied:
The reason why I split this up was because the error message was off for the second case:
1. Error message makes sense here because we're installing something: https://github.com/purescript/spago/pull/1018/files#diff-4529e47b55cd312f9954aa4027ac032176d4c22848d7e5ccf2c7adb942587e38L99
2. but not here because we could NOT pass the `install` flag and instead just be doing the ensure ranges. In that situation, why would the error mention anything about installation? https://github.com/purescript/spago/pull/1018/files#diff-4529e47b55cd312f9954aa4027ac032176d4c22848d7e5ccf2c7adb942587e38L112

Oh, and where is the third time it's being duplicated?

f-f replied:
Lines 104 and 128 are the same messages with different structures, 146 is a slightly different but with the same structure as 128.

I guess this is alright, but we should at least collect the shared error message in one place

JordanMartinez replied:
Hm.... I looked at sharing the error message, but then came up with a different idea. What if we defined the current package install block of code and 'ensureRanges' block of code into let-bound functions and then just call those functions in their respective places?

f-f replied:
I'm not sure how that would look like, but we can try it out 😄

JordanMartinez replied:
I think there's a bug in the current PR. If I have a package in the workspace `spago.yaml`, and don't select a package BUT I do have the `ensureRanges` flag, then the the root package doesn't get ranges added to it. I believe this occurs because I only mark the root package as a `SinglePackage` if the `packages` to install is not null. However, I don't also check whether the `ensureRanges` value is true.

f-f replied:
Ah I see - can we write a test to verify this?

JordanMartinez replied:
I still need to add a test, but see latest commits for what I did here.

JordanMartinez replied:
Hm... This still seems to differ from the current behavior. Sigh...

JordanMartinez replied:
I've updated the code back to the original approach here. Seems simpler at the end of the day.

f-f commented on src/Spago/Command/Build.purs:
```diff
@@ -140,56 +96,188 @@ run opts = do
+          unless (Array.null errors) do
+            die' errors
+
+  when buildingMultiplePackages do
+    moduleMap <- liftEffect $ Ref.read duplicateModuleRef
+    let duplicateModules = Map.filter (\packages -> Array.length packages > 1) moduleMap
+    unless (Map.isEmpty duplicateModules) do
+      die $
+        ( Dodo.lines
+            [ Dodo.text $ "Detected " <> show (Map.size duplicateModules) <> " modules with the same module name across 2 or more packages defined in this workspace."
+            , duplicateModules
+                # Map.toUnfoldable
+                # Array.mapWithIndex
+                    ( \idx (Tuple m ps) ->
+                        Dodo.lines
+                          [ Dodo.text $ show (idx + 1) <> ") Module \"" <> m <> "\" was defined in the following packages:"
+                          , Dodo.indent
+                              $ Dodo.lines
+                              $ map
+                                  ( \(Tuple pkg moduleFilePath) ->
+                                      Dodo.text $ "- " <> PackageName.print pkg <> "   at path: " <> moduleFilePath
+                                  )
+                              $ Array.sort ps
+                          ]
+                    )
+                # Dodo.lines
+            ] :: Docc
+        )
 
```
Let's extract this formatting stuff to a separate function

JordanMartinez replied:
Done.

f-f commented on src/Spago/Command/Build.purs:
```diff
@@ -140,56 +96,188 @@ run opts = do
   workspacePackageGlob :: WorkspacePackage -> Array String
   workspacePackageGlob p = Config.sourceGlob testGlobs p.package.name (WorkspacePackage p)
 
+  { projectGlobs, monorepoTestGlobs } = case depsOnly of
```
There was a comment detailing the reason why the depsOnly works that way, let's reintroduce it here

f-f commented on test/Spago/Build/Polyrepo.purs:
```diff
@@ -0,0 +1,349 @@
+spec = Spec.describe "polyrepo" do
+
+  let
+    spagoInitCleanupNonPackageFiles spago = do
+      spago [ "init" ] >>= shouldBeSuccess
+      FSA.unlink $ Path.concat [ "src", "Main.purs" ]
+      FSA.unlink $ Path.concat [ "test", "Test", "Main.purs" ]
+      FSA.unlink "spago.yaml"
+      FS.writeYamlFile Config.configCodec "spago.yaml"
+        $ Init.defaultConfig'
+        $ WorkspaceOnly { setVersion: Just $ unsafeFromRight $ Version.parse "0.0.1" }
```
I think this could just be this?

```suggestion
    spagoInitConfig spago = do
      FS.writeYamlFile Config.configCodec "spago.yaml"
        $ Init.defaultConfig'
        $ WorkspaceOnly { setVersion: Just $ unsafeFromRight $ Version.parse "0.0.1" }
```

f-f replied:
I don't understand - `spago init` boils down to "make a Main, a Test.Main, and a basic spago.yaml". (there are a few more details, but they are negligible)

In the snippet above you call `spago init`, delete the three main things it creates, and then make a new configuration.
Hence why I'm saying: we might as well just make the configuration without calling the `init` first?

JordanMartinez replied:
Ah, I see your point now.

JordanMartinez replied:
Done.

f-f commented on test/Spago/Build/Polyrepo.purs:
```diff
@@ -0,0 +1,349 @@
+import Test.Spec (SpecT)
+import Test.Spec as Spec
+import Test.Spec.Assertions as Assert
+
```
Might be worth linking this PR here in this module, so that visitors from the future will be able to find the diagrams in the PR description, as understanding the various cases with those is easier

JordanMartinez replied:
What happens if we add more tests here? Do we link to those PRs as well?

I think I'd rather copy-paste the Mermaid diagram into this module's docs. If we add a new test to this situation, we can update the diagrams as well.

f-f replied:
Ah yes good point. Let's go with adding the mermaid diagrams here.

f-f reviewed (commented):
Finally managed to go through the whole thing. Good work @JordanMartinez! I left a few small notes but it's otherwise good to go

f-f commented:
> I think this should be further discussed before we work on it. At the very least, I don't plan on doing anything for this in this PR.

Ah yes about this: agreed that we can shrink the PSA display in a future PR, but I'd like to address the multiple `Build Succeeded` messages in here, as they don't really make sense right now

JordanMartinez commented:
> > I think this should be further discussed before we work on it. At the very least, I don't plan on doing anything for this in this PR.
> 
> Ah yes about this: agreed that we can shrink the PSA display in a future PR, but I'd like to address the multiple `Build Succeeded` messages in here, as they don't really make sense right now

What about them doesn't make sense? I don't have enough of an idea for how to move forward on this.

f-f commented:
> What about them doesn't make sense? I don't have enough of an idea for how to move forward on this.

The `build` is - even if split into multiple parts as we are doing here - a single process with a boolean outcome: either it's good and you can move on, or it's broken and needs fixing.

The problem with multiple "Build succeeded" messages is that it breaks this perspective, a build can now succeed multiple times, but fail only once.

To tackle this we should move this snippet out of the `Psa` module and into the `Build` module:
https://github.com/purescript/spago/blob/899b4939b782aa04932a4c038e72479b9cc8705c/src/Spago/Psa.purs#L109-L113

We'd return this list of errors from the Psa command instead of killing the process there, and draw the final verdict on the build only once they all succeed or one of them fails.

JordanMartinez commented:
> We'd return this list of errors from the Psa command instead of killing the process there, and draw the final verdict on the build only once they all succeed or one of them fails.

How does this work when a non-js backend is used? There's already a "Backend build succeeded" message.

JordanMartinez commented:
Remaining tasks:
- the `ensureRanges` test I have yet to add
- address feedback on latest changes
- address the psa `Build Succeeded` issue

f-f commented:
> How does this work when a non-js backend is used? There's already a "Backend build succeeded" message.

Ah yes, indeed - when building multiple packages with a backend we'd first build with purs, then build with the backend, then repeat. Alright, let's leave this out of this PR and just open an issue once we merge

f-f reviewed (approved):
This is great now - thanks @JordanMartinez!

Let's add the last test for `ensureRanges`, then we can merge in

JordanMartinez commented:
In latest commits, I...
- moved the build-in-topological-order tests into their own `describe` block
- kept case 4 (duplicate module check) in the same location, but renamed it for clarity
- moved Mermaid diagrams above each test for readability, and linked to the guide for how to write one for future users
- added `--ensure-ranges` tests:
   - poly repo where a root package does and does not exist
   - monorepo where a root package exists. I didn't add one for when a root package does not exist as I don't think this will ever happen.
   
I won't merge this despite the approval in case you have more feedback. But if it looks good, then feel free to merge it.

JordanMartinez commented:
> > How does this work when a non-js backend is used? There's already a "Backend build succeeded" message.
> 
> Ah yes, indeed - when building multiple packages with a backend we'd first build with purs, then build with the backend, then repeat. Alright, let's leave this out of this PR and just open an issue once we merge

I've tracked our conversation about this here and added it to the original issue: https://github.com/purescript/spago/issues/988#issuecomment-1734388070
